src/components/EmailValidationForm.tsx

```diff
@@ -1,71 +1,88 @@
 
 export interface EmailValidationFormState {
-  sent: boolean;
+  validated: boolean;
 }
 
 export interface EmailValidationFormOwnProps {
-  library: LibraryData;
+  email: string;
+  libraryInfo: {};
+  uuid: string;
   store: Store<State>;
-  fetchLibrary: (uuid: string) => LibraryData;
-}
-
-export interface EmailValidationFormDispatchProps {
-  validateEmail: (data: FormData) => Promise<void>;
 }
 
 export interface EmailValidationFormStateProps {
   error?: FetchErrorData;
 }
 
+export interface EmailValidationFormDispatchProps {
+  validateEmail: (data: FormData) => Promise<void>;
+}
+
 export interface EmailValidationFormProps extends EmailValidationFormOwnProps, EmailValidationFormDispatchProps, EmailValidationFormStateProps {}
 
 export class EmailValidationForm extends React.Component<EmailValidationFormProps, EmailValidationFormState> {
-
   constructor(props) {
     super(props);
-    this.sendEmail = this.sendEmail.bind(this);
-    this.state = { sent: false };
+    this.validate = this.validate.bind(this);
+    this.renderInfoText = this.renderInfoText.bind(this);
+    this.state = { validated: false };
   }
 
-  async sendEmail(data: FormData): Promise<void> {
-    await this.props.validateEmail(data);
-    await this.props.fetchLibrary(this.props.library.uuid);
-    this.setState({ sent: !this.props.error });
+  convert(email: string): string {
+    return (email.includes("_") ? email.replace("_", " ") : email.replace(" ", "_"));
   }
 
-  render(): JSX.Element {
-    let icon = this.state.sent ? <CheckSoloIcon /> : null;
-    let alreadyValidated = this.props.library.urls_and_contact.validated !== "Not validated";
-    let hasEmail = !!this.props.library.urls_and_contact.contact_email;
+  renderTitle(emailString: string, emailAddress: string, alreadyValidated: boolean): JSX.Element[] {
+    let titleString = `${emailString}${emailAddress ? ": " + emailAddress : ""}`;
+    let icon = ((alreadyValidated || this.state.validated) && !this.props.error) ? <CheckSoloIcon /> : <XIcon />;
+    return [<span>{titleString}</span>, icon];
+  }
+
+  async validate(email: string, data: FormData): Promise<void> {
+    data.append("email", email);
+    await this.props.validateEmail(data);
+    this.setState({ validated: true });
+  }
 
-    let buttonText = hasEmail ? "Validate email address" : "No email address configured";
-    let buttonContent = <span>{buttonText}{icon}</span>;
-    let infoText = (alreadyValidated && !this.state.sent && !this.props.error) ? "Already validated" : null;
-    let successText = `Successfully validated ${this.props.library.urls_and_contact.contact_email}`;
+  renderInfoText(emailString: string, emailAddress: string, validated: string, alreadyValidated: boolean): string {
+    if (!emailAddress) {
+      return `No ${emailString} configured`;
+    } else if (!alreadyValidated) {
+      return "Not validated";
+    } else if (!this.state.validated && !this.props.error) {
```

Comment:
Is this right? If it's not validated and there's no error, then display "validated"?

Also, what if there is an email, it's not already validated, and this last `else if` fails? There's no default return case.

Reply:
Hmmm this makes sense because if `Form`'s `infoText` is undefined, nothing gets rendered, but if you didn't already know that or if `Form` gets updated in the future, you may want to be explicit about what is returned so there's no error.

Reply:
Yes, that's right.  `this.state.validated` isn't whether the email has previously been validated, it's whether you've clicked "validate" just now.  So:
-- If it's never been validated, it displays "not validated".
-- If it has been validated at some point in the past, but not by you having clicked it just now, it displays "validated".
-- If it's validated because you just clicked it, i.e. if `this.state.validated` is true, it displays "successfully validated" and turns green (that's the `successText` being passed to the Form).

Reply:
So, if you just clicked to validate it, `this.state.validated` becomes true and the `successText` is shown, that I get. But this last `else if` is checking that `this.state.validated` is not true and there's no error. Is it assuming that it was therefore already validated at `validated` time stamp?

Reply:
Yes--that's the only other possibility.  We've already covered "there's no email address" and "there's an email address but it's never been validated."